haystack/nodes/other/shaper.py

```diff
@@ -0,0 +1,419 @@
+                    f"an assignment. The input configuration is {node}"
+                )
+
+    def invoke_function(self, parent: str, node: Dict[str, Any], invocation_context: Dict[str, Any]) -> None:
+        """
+        Invoke a function on the given invocation context and stores the result back in the invocation context.
+
+        :param parent: The parent node of the current node.
+        :param node: A dictionary representing the function node.
+        :param invocation_context: A dictionary of variables available in the invocation context.
+        :raises ValueError: If the parameters specified in the YAML definition of the function are invalid.
+        """
+        params = node.get("params", [])
+        output = node.get("output", parent)
+        # use case when the function is invoked with positional arguments
+        if isinstance(params, list):
+            # resolve the argument values from the invocation context
+            args = [invocation_context[param] for param in params if param in invocation_context]
+
+            if len(args) != len(params):
+                raise ValueError(
+                    f"Invalid function arguments. The function {node['func']} "
+                    f"specifies in its YAML definition {len(params)} arguments yet {len(args)} arguments "
+                    f"were resolved"
+                )
+            # push the output variable name to the invocation stack
+            if output not in invocation_context:
+                invocation_context["invocation_stack"].append(output)
+            # and finally invoke the function with the resolved arguments
+            invocation_result = self.invoke_function_with_args(parent, node, invocation_context, *args)
+
+        # use case when the function is invoked with keyword arguments
+        elif isinstance(params, dict):
+            kwargs = {}
+            for key, value in params.items():
+                resolved_value = invocation_context.get(key)
+                if resolved_value:
+                    # if the key is a variable available in the invocation context, resolve it
+                    kwargs[key] = resolved_value
+                else:
+                    # attempt to find the value from the invocation context and assign to the keyword argument
+                    kwargs[key] = value if value not in invocation_context else invocation_context[value]
+                if key in invocation_context and invocation_context["invocation_stack"][-1] == key:
+                    invocation_context["invocation_stack"].pop()
+                    invocation_context.pop(key)
+            invocation_result = self.invoke_function_with_kwargs(parent, node, invocation_context, **kwargs)
+        else:
+            raise ValueError(f"Invalid YAML definition {node}")
+
+        # finally store the result in the invocation context
+        invocation_context[output] = invocation_result
```

Comment:
I think the `Shaper` would be even more useful, if we could also re-shape actual pipeline kwargs. I understand that we can't pass arbitrary kwargs but if I want to re-shape documents, I could easily do it (i.e. instead of adding the reshaped documents to `invocation_context["documents"]` replace `documents` directly. It would tie the node less to the `PromptNode`. Maybe add a parameter wether we want to override actual pipeline kwargs (if they exist) or if it should be added to `invocation_context`?

Reply:
An example use-case would be:

I want to perform query expansion using a `PromptNode` and pass the concatenated query to a `BM25Retriever` afterwards. Let's say I want to generate 10 additional questions per query. The way I would do it:

```
Query >> PromptNode >> generate 10 questions (can we define top-k for the PromptNode?) >> Shaper >> concatenate all generated questions and original query into Query >> BM25Retriever >> return Documents
```

Reply:
Yeah we can do this @mathislucka

Reply:
@ZanSara can correct me but I think we can take a variable from invocation context or a normal variable, transform it and use it as one of the familiar variables e.g. documents